### PR TITLE
[filesystem] Fixes #4754 #4248 Parse content-disposition properly and fix firefox download

### LINKED ISSUE #4248: Download menu entry doesn't work on Firefox
Steps to reproduce:

1. Open [Theia in Theia (Gitpod.io)](https://gitpod.io/#github.com/theia-ide/theia)
2. Right-click on any file in the file tree, then click the `Download` menu entry

Expected behavior:

- The file gets downloaded

Actual behavior:

- Nothing happens (reproduced in both Firefox and Chrome, and on both Linux and Mac)

### LINKED ISSUE #4754: [Filesystem] 500 Error when downloading files with non-ascii characters
### Description
Theia does not allow downloading of files/folders that uses some non-ascii characters on their filenames/folder names.

### Reproduction Steps
- Create a new file and name it `Φtest.txt` (you can copy/paste that and use as the filename)
- Once created download the file, Right-click -> download. It will not do anything, and if you check your browser's console, you will see an error which is a `500 error`

Screenshot:
image

**OS and Theia version:**
Win10 - Theia Master branch

### PR DESCRIPTION
**Pull request by uniibu**
This is in continuation of PR #4755.

This commit Fixes #4754 and fixes #4248. It uses express' own attachment method and parses the content-disposition reliably and decodes the filename if possible. 

This also fixes the downloading on firefox by appending the anchor link to the body before triggering the click event and removes it after.

Signed-off-by: Uni Sayo <unibtc@gmail.com>

### CODE REVIEW

**Comment by akosyakov** [comment]
@kittaakos Could you please have a look at it? Does it fix https://github.com/gitpod-io/gitpod/issues/414?

**Comment by kittaakos** [comment]
From https://github.com/theia-ide/theia/pull/4755#issuecomment-477965965:

> @kittaakos The problem with encodeURIComponent on Content-Disposition is browser compatibility. It works fine on Chrome, but like in Firefox it has to be set as filename*="utf-8''${encodeURIComponent(name)}" example and there are also other incompatibilities on other browsers such as safari and ie when simply using encodeURIComponent. Another filename you could test are with chinese characters like 中指南.txt

**Comment by kittaakos** [comment]
> Another filename you could test are with chinese characters like 中指南.txt

I could download both `中指南.txt` and `Φtest.txt` applying only the [encoding/decoding part](https://github.com/theia-ide/theia/pull/4755#pullrequestreview-220421683).

I am checking your [DOM related changes](https://github.com/theia-ide/theia/pull/4762/files#diff-37b434f1c869972763530f692f0b2fd8R201) in FF.

**Comment by kittaakos** [comment]
> I am checking your [DOM related changes](https://github.com/theia-ide/theia/pull/4762/files#diff-37b434f1c869972763530f692f0b2fd8R201) in FF.

Yes, we need your DOM related change. Very nice 👍. Also confirmed here: https://support.mozilla.org/en-US/questions/968992


Long story short; downloading of non-ASCII files work in both FF and Chrome with this bare minimum change on my side. I am bit relactant to introduce your disposition parsing logic unless you provide me a use-case that works only with your propsed changes:

```diff
diff --git a/packages/filesystem/src/browser/download/file-download-service.ts b/packages/filesystem/src/browser/download/file-download-service.ts
index b870a97ca..075f54126 100644
--- a/packages/filesystem/src/browser/download/file-download-service.ts
+++ b/packages/filesystem/src/browser/download/file-download-service.ts
@@ -194,12 +194,17 @@ export class FileDownloadService {
             url = URL.createObjectURL(blob);
             if (this.anchor === undefined) {
                 this.anchor = document.createElement('a');
-                this.anchor.style.display = 'none';
             }
             this.anchor.href = url;
             this.anchor.download = title;
+            this.anchor.style.display = 'none';
+            document.body.appendChild(this.anchor);
             this.anchor.click();
         } finally {
+            // make sure anchor is removed from parent
+            if (this.anchor && this.anchor.parentNode) {
+                this.anchor.parentNode.removeChild(this.anchor);
+            }
             if (url) {
                 URL.revokeObjectURL(url);
             }
@@ -209,7 +214,7 @@ export class FileDownloadService {
     protected async title(response: Response, uris: URI[]): Promise<string> {
         let title = (response.headers.get('Content-Disposition') || '').split('attachment; filename=').pop();
         if (title) {
-            return title;
+            return decodeURIComponent(title);
         }
         // tslint:disable-next-line:whitespace
         const [uri,] = uris;
diff --git a/packages/filesystem/src/node/download/file-download-handler.ts b/packages/filesystem/src/node/download/file-download-handler.ts
index 52888e49f..f812b6df7 100644
--- a/packages/filesystem/src/node/download/file-download-handler.ts
+++ b/packages/filesystem/src/node/download/file-download-handler.ts
@@ -53,7 +53,7 @@ export abstract class FileDownloadHandler {
         } else {
             this.logger.debug(`Cannot determine the content-type for file: ${filePath}. Skipping the 'Content-type' header from the HTTP response.`);
         }
-        response.setHeader('Content-Disposition', `attachment; filename=${name}`);
+        response.setHeader('Content-Disposition', `attachment; filename=${encodeURIComponent(name)}`);
         try {
             await fs.access(filePath, fs.constants.R_OK);
             fs.readFile(filePath, (error, data) => {
```

**Comment by uniibu** [comment]
I'm a bit hesitant if we should keep this kind of download (using fetch). This will limit the user from download big files and also prone to out of memory error since the data is kept in memory until it completes.

I think instead of fetching it, why not directly put it on the anchor link as the `href` and just add an extra query to indicate if the file needs to be deleted.

The flow would be like:
- [frontend] sends the uri/s to the backend
- [backend] processes the uri/s archive them if applicable, then respond with a simple json payload that includes the filepath, the filename to download and if this file needs to be deleted.
- [frontend] constructs the href of the anchor link based from the json payload and initiate the download.
- [backend] receives the download request and streams the download to the user, and after streaming, it checks the query if the file needs to be deleted.
This way it will not limit the user from downloading big files.

**Comment by dickroberts** [comment]
thanks for the fix!

**Comment by uniibu** [comment]
That was weird.. Made a comment and then delete his account o.o

**Comment by rickiebobby** [comment]
how do i pull this fix into package.json?

**Comment by kittaakos** [comment]
@uniibu, freely speaking; it is a bit hard to follow you. What do you fix? First, you came up with the ASCII filename issue, then the FF download problem, now with the file size...

**Comment by kittaakos** [comment]
> @kittaakos Could you please have a look at it? Does it fix [gitpod-io/gitpod#414](https://github.com/gitpod-io/gitpod/issues/414)?

I have checked it locally; I could download a ~2 GB iso file, but it worked locally with the HEAD (237e6b2c4d956c9367e1fc92c87529855e4d9e79) of the `master` too.

@akosyakov, how can I try this with a real remote setup? Do you know a GH repo with a gigantic file resource? Thanks!

**Comment by kittaakos** [comment]
When I tried to download a file larger than 4 GB, I got the same error with this change and the `master` too, I doubt it fixes the issue:

```
oot ERROR Error occurred when downloading: file:///path/to/your/file. Error: Received unexpected status code: 500. [Internal Server Error]
    at FileDownloadService.<anonymous> (http://localhost:3000/6.bundle.js:583:35)
    at step (http://localhost:3000/6.bundle.js:408:23)
    at Object.next (http://localhost:3000/6.bundle.js:389:53)
    at fulfilled (http://localhost:3000/6.bundle.js:380:58)
log @ logger-protocol.ts:112
(anonymous) @ logger-frontend-module.ts:41
(anonymous) @ logger.ts:312
(anonymous) @ logger.ts:304
Promise.then (async)
../../packages/core/lib/common/logger.js.Logger.log @ logger.ts:299
../../packages/core/lib/common/logger.js.Logger.error @ logger.ts:370
(anonymous) @ file-download-service.ts:176
step @ file-download-service.ts:15
(anonymous) @ file-download-service.ts:15
fulfilled @ file-download-service.ts:15
Promise.then (async)
step @ file-download-service.ts:15
fulfilled @ file-download-service.ts:15
Promise.then (async)
step @ file-download-service.ts:15
fulfilled @ file-download-service.ts:15
Promise.then (async)
step @ file-download-service.ts:15
fulfilled @ file-download-service.ts:15
Promise.then (async)
step @ file-download-service.ts:15
(anonymous) @ file-download-service.ts:15
push.../../packages/filesystem/lib/browser/download/file-download-service.js.__awaiter @ file-download-service.ts:15
push.../../packages/filesystem/lib/browser/download/file-download-service.js.FileDownloadService.download @ file-download-service.ts:150
(anonymous) @ file-download-command-contribution.ts:75
step @ file-download-command-contribution.ts:15
(anonymous) @ file-download-command-contribution.ts:15
(anonymous) @ file-download-command-contribution.ts:15
push.../../packages/filesystem/lib/browser/download/file-download-command-contribution.js.__awaiter @ file-download-command-contribution.ts:15
push.../../packages/filesystem/lib/browser/download/file-download-command-contribution.js.FileDownloadCommandContribution.executeDownload @ file-download-command-contribution.ts:74
execute @ file-download-command-contribution.ts:68
../../packages/core/lib/common/uri-command-handler.js.UriAwareCommandHandler.execute @ uri-command-handler.ts:89
(anonymous) @ command.ts:241
step @ command.ts:15
(anonymous) @ command.ts:15
(anonymous) @ command.ts:15
../../packages/core/lib/common/command.js.__awaiter @ command.ts:15
../../packages/core/lib/common/command.js.CommandRegistry.executeCommand @ command.ts:238
execute @ browser-menu-plugin.ts:96
../../node_modules/@phosphor/commands/lib/index.js.CommandRegistry.execute @ index.js:351
../../node_modules/@phosphor/widgets/lib/menu.js.Menu.triggerActiveItem @ menu.js:305
../../node_modules/@phosphor/widgets/lib/menu.js.Menu._evtMouseUp @ menu.js:641
../../node_modules/@phosphor/widgets/lib/menu.js.Menu.handleEvent @ menu.js:451
```

**Comment by uniibu** [comment]
> When I tried to download a file larger than 4 GB, I got the same error with this change and the `master` too, I doubt it fixes the issue:
> 
Yeah the issue is that on backend it uses `fs.readFile` instead of `fs.createReadStream` and and if ever it was able to pass tru the backend without memory issue, you'll end up with memory issue on the frontend as well because of the fetch. 

Maybe i can just have this pr fix the the encoding and the firefox issue, then just discuss the rest on another.

**Comment by uniibu** [comment]
@kittaakos i've reverted some changes and simply use encodeURIComponent and decodeURIComponent with the inclusion of the fix for the firefox download. Commits squashed. 

_-This does not fix downloading of big files.-_

**Comment by kittaakos** [comment]
> @kittaakos i've reverted some changes and simply use encodeURIComponent and decodeURIComponent with the inclusion of the fix for the firefox download. Commits squashed.

Thanks, I am trying it again.

**Comment by kittaakos** [comment]
@akosyakov, are you OK merging this? This patch solves the non-ASCII filename problem and the download issue in FF.

**Comment by akosyakov** [comment]
@kittaakos :+1
